### ISSUE TITLE
Extremely long compilation times with Batch Normlaization

### ISSUE THREAD

**Issue by botev** [Theano/Theano#4741]
I'm getting extremely long times for a Theano function to compile with Batch Normalization when using the GPU (with cuDNN v5). I'm using the bleeding edge version, I'm also using `conv_meta` optimizer. Without BN the compilation time is 3-5min, at the moment it has been half an hour since I launched it on the server and is still in compilation phase. Any ideas?

**Comment by nouiz** [comment]
Do you use Theano batch norm or the very recently wrapped cudnn batch norm?

the cudnn batch norm make the compilation faster (as the graph is smaller):

http://deeplearning.net/software/theano_versions/dev/library/sandbox/cuda/dnn.html#batch-normalization

Otherwise, I'm investigating a graph that have super-exponential time in
the size of the graph. After some size just one opt cause super slow
compilation time. You can disable just that optimization and see if this is
your case too. Use this flag:

optimizer_excluding=local_greedy_distributor

Otherwise, can you make your graph have less layers so that it compile
under 1h and run Theano compilation profiler and send me the output? use
the flag profile=True,profile_optimizer=True.

**Comment by botev** [comment]
I have no problem with compiling the whole graph for an hour, since its on a server. However, are you sure you want the whole graph and in which format do you want me to send it to you? (the print or a picture). Also should I just post it here?

**Comment by nouiz** [comment]
Redirect the output to a file and send this.

Just to be sure I was clear, if the compilation take more then 1h, you
should make the graph smaller or let it run. Killing the compilation
process won't give me the information I need.

**Comment by nouiz** [comment]
no news and the cudnn bn should make the compilation fast, so closing.
